src/Data/Text/Lazy.hs

```diff
@@ -794,12 +794,10 @@ foldr1 f t = S.foldr1 f (stream t)
 
 -- | /O(n)/ Concatenate a list of 'Text's.
 concat :: [Text] -> Text
-concat = to
-  where
-    go Empty        css = to css
-    go (Chunk c cs) css = Chunk c (go cs css)
-    to []               = Empty
-    to (cs:css)         = go cs css
+concat []                    = Empty
+concat (Empty : css)         = concat css
+concat (Chunk c Empty : css) = Chunk c (concat css)
+concat (Chunk c cs : css)    = Chunk c (concat (cs : css))
```

Comment:
Why this redesign of `concat`? Compared to the old version, it looks like it will allocate an extra cons cell on this line here.

Reply:
Agree here - this doesn't seem like it needed a rewrite unless there was some specific optimization in mind. I don't see anything off the top of my head that would improve laziness or matter that much in terms of optimizations tho.

Reply:
The old version had two mutually recursive functions, so was not always tail recursive.  In particular concat of a long list of `Empty` could perhaps run out of stack space? The new code should handle that. 

I'd be tempted to change the order of cases to optimise for the non-empty case:
```haskell
concat (Chunk c Empty : css) = Chunk c (concat css)
concat (Chunk c cs : css)    = Chunk c (concat (cs : css))
concat (Empty : css)         = concat css
concat []                    = Empty
```

And then, if it is in fact cheaper to allocate a closure (thunk) around two values, rather than a closure around a cons-cell, avoid the cons cell allocation via something along the lines of:

```haskell
concat (Chunk c Empty : css)   = Chunk c (concat css)
concat (Chunk c cs : css)      = Chunk c (concat2 cs css) where
  concat2 (Chunk c Empty) css  = Chunk c (concat css)
  concat2 (Chunk c cs) css     = Chunk c (concat2 cs css)
  concat2 Empty (cs : css)     = concat2 cs css
  concat2 Empty _              = Empty
concat (Empty : css)           = concat css
concat []                      = Empty
```

Reply:
@gwils SpecConstr transformation takes care to eliminate allocation of a cons cell, so there is no inefficiency introduced. The purpose, as noted by @vdukhovni, is indeed to be more tail recursive. It helps with a list of `Empty`, but also improves performance of a common pattern when concatenating a list of strict `Text` into a lazy one: now we stay within a single function most of the time. Here is a slightly edited Core dump:

```haskell
concat :: [Text] -> Text
concat ds_d7dL = case ds_d7dL of
  [] -> Empty
  : ds1_d7fW css_a7c4 -> case ds1_d7fW of
    Empty -> concat css_a7c4;
    Chunk dt_d7g0 dt1_d7g1 dt2_d7g2 ds2_d7fX -> case ds2_d7fX of
      Empty -> Chunk dt_d7g0 dt1_d7g1 dt2_d7g2 (concat css_a7c4)
      Chunk ipv_s7g7 ipv1_s7g8 ipv2_s7g9 ipv3_s7ga ->
        Chunk dt_d7g0 dt1_d7g1 dt2_d7g2
          (concat_$sconcat ipv_s7g7 ipv1_s7g8 ipv2_s7g9 ipv3_s7ga css_a7c4)

concat_$sconcat :: ByteArray# -> Int# -> Int# -> Text -> [Text] -> Text
concat_$sconcat sc_s7hC sc1_s7hD sc2_s7hE sc3_s7hF sc4_s7hG = case sc3_s7hF of
  Empty -> Chunk sc_s7hC sc1_s7hD sc2_s7hE (concat sc4_s7hG)
  Chunk ipv_s7g7 ipv1_s7g8 ipv2_s7g9 ipv3_s7ga ->
    Chunk sc_s7hC sc1_s7hD sc2_s7hE
      (concat_$sconcat ipv_s7g7 ipv1_s7g8 ipv2_s7g9 ipv3_s7ga sc4_s7hG)
```

@vdukhovni AFAIK order of cases does not matter (order of constructors does).